File: hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/hap/entity/HwSignHead.java
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.hap.entity;

import com.ohos.hapsigntool.utils.ByteArrayUtils;

import java.io.IOException;

/**
 * define class of hap signature block head
 *
 * @since 2021-12-13
 */
public class HwSignHead {
    /**
     * length of sign head
     */
    public static final int SIGN_HEAD_LEN = 32;

    private static final char[] MAGIC = "hw signed app   ".toCharArray(); // 16Bytes-Magic
    private static final char[] VERSION = "1000".toCharArray(); // 4-Bytes, version is 1.0.0.0
    private static final int NUM_OF_BLOCK = 2; // number of sub-block
    private static final int RESERVE_LENGTH = 4;
    private char[] reserve = new char[RESERVE_LENGTH];

    /**
     * get serialization of HwSignHead
     *
     * @param subBlockSize the total size of all sub-blocks
     * @return Byte array after serialization of HwSignHead
     */
    public byte[] getSignHead(int subBlockSize) {
        int size = subBlockSize; // total size of sub-block
        byte[] signHead = new byte[SIGN_HEAD_LEN];
        int start = 0;
        try {
            start = ByteArrayUtils.insertCharToByteArray(signHead, start, MAGIC);
            if (start < 0) {
                throw new IOException();
            }
            start = ByteArrayUtils.insertCharToByteArray(signHead, start, VERSION);
            if (start < 0) {
                throw new IOException();
            }
            start = ByteArrayUtils.insertIntToByteArray(signHead, start, size);
            if (start < 0) {
                throw new IOException();
            }
            start = ByteArrayUtils.insertIntToByteArray(signHead, start, NUM_OF_BLOCK);
            if (start < 0) {
                throw new IOException();
            }
            start = ByteArrayUtils.insertCharToByteArray(signHead, start, reserve);
            if (start < 0) {
                throw new IOException();
            }
        } catch (IOException e) {
            return null;
        }
        return signHead;
    }
}

File: pc/services/hap/entity/src/hw_sign_head.cpp
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include "byte_buffer.h"
#include "hw_sign_head.h"
#include "byte_array_utils.h"
#include "signature_tools_log.h"

using namespace OHOS::SignatureTools;

const int HwSignHead::SIGN_HEAD_LEN = 32;
const std::string HwSignHead::MAGIC = "hw signed app   ";
const std::string HwSignHead::ELF_MAGIC = "elf sign block  ";
const std::string HwSignHead::VERSION = "1000";
const int HwSignHead::NUM_OF_BLOCK = 2;
const int HwSignHead::RESERVE_LENGTH = 4;
const int32_t HwSignHead::ELF_BLOCK_LEN = 12;
const int32_t HwSignHead::BIN_BLOCK_LEN = 8;
std::vector<int8_t> HwSignHead::reserve = std::vector<int8_t>(HwSignHead::RESERVE_LENGTH, 0);

HwSignHead::HwSignHead() { 

}

std::vector<int8_t> HwSignHead::GetSignHead(int subBlockSize)
{
    int size = subBlockSize;
    std::vector<int8_t> signHead(SIGN_HEAD_LEN, 0);
    int start = 0;
    start = ByteArrayUtils::InsertCharToByteArray(signHead, start, MAGIC);
    if (start < 0) {
        SIGNATURE_TOOLS_LOGW("InsertCharToByteArray failed.\n");
        return std::vector<int8_t>();
    }
    start = ByteArrayUtils::InsertCharToByteArray(signHead, start, VERSION);
    if (start < 0) {
        SIGNATURE_TOOLS_LOGW("InsertCharToByteArray failed.\n");
        return std::vector<int8_t>();
    }
    start = ByteArrayUtils::InsertIntToByteArray(signHead, start, size);
    if (start < 0) {
        SIGNATURE_TOOLS_LOGW("InsertIntToByteArray failed.\n");
        return std::vector<int8_t>();
    }
    start = ByteArrayUtils::InsertIntToByteArray(signHead, start, NUM_OF_BLOCK);
    if (start < 0) {
        SIGNATURE_TOOLS_LOGW("InsertIntToByteArray failed.\n");
        return std::vector<int8_t>();
    }
    start = ByteArrayUtils::InsertCharToByteArray(signHead, start, std::string(reserve.begin(), reserve.end()));
    if (start < 0) {
        SIGNATURE_TOOLS_LOGW("InsertCharToByteArray failed.\n");
        return std::vector<int8_t>();
    }
    return signHead;
}
std::vector<int8_t> HwSignHead::getSignHeadLittleEndian(int subBlockSize, int subBlockNum)
{
    ByteBuffer bf = ByteBuffer(HwSignHead::SIGN_HEAD_LEN);
    for (char c : HwSignHead::ELF_MAGIC)
    {
        bf.PutByte(c);
    }
    for (char c : HwSignHead::VERSION)
    {
        bf.PutByte(c);
    }
    bf.PutInt32(subBlockSize);
    bf.PutInt32(subBlockNum);
    for (char c : HwSignHead::reserve)
    {
        bf.PutByte(c);
    }
    int8_t ret[HwSignHead::SIGN_HEAD_LEN];
    bf.GetData(0, (char *)&ret, HwSignHead::SIGN_HEAD_LEN);
    std::vector<int8_t> byte(ret, ret + HwSignHead::SIGN_HEAD_LEN);

    return byte;
}